### PR TITLE
add concat function dtype test

### LINKED ISSUE #32934: Incorrect dtype for concat with empty float and datetime columns
#### Code Sample, a copy-pastable example if possible

```python
In [40]: a = pd.DataFrame({"A": pd.array(['2000'], dtype='datetime64[ns]')})

In [41]: b = pd.DataFrame({"A": pd.array([1.0], dtype='float64')})

In [42]: pd.concat([a.iloc[:0], b.iloc[:0]])['A'].dtype
Out[42]: dtype('<M8[ns]')
```

#### Problem description

The return dtype of concatentating a float64 and datetime64[ns] should be object dtype. That's typically the case when both are non-empty

```python
In [45]: pd.concat([a, b])['A'].dtype
Out[45]: dtype('O')
```

But when either is empty, we get incorrect results

```python
In [42]: pd.concat([a.iloc[:0], b.iloc[:0]])['A'].dtype
Out[42]: dtype('<M8[ns]')

In [43]: pd.concat([a.iloc[:0], b])['A'].dtype
Out[43]: dtype('float64')

In [44]: pd.concat([a, b.iloc[:0]])['A'].dtype
Out[44]: dtype('<M8[ns]')
```

All of those should be object dtype.

### PR DESCRIPTION
**Pull request by liang3zy22**
- [ ] closes #32934
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

The old PR closed, so created new one to close #32934.

### CODE REVIEW

**Comment by liang3zy22** [comment]
1 check failed with current code,Ubuntu/Data Manager. It means that the code
`pd.concat([df_time, df_float.iloc[:0]])`
has data type "object".
If I updated the codes to the following code:
` tm.assert_equal(
        pd.concat([df_time, df_float.iloc[:0]]),
        pd.DataFrame({"A": pd.array(["2000"], dtype="datetime64[ns]")}).astype({"A":"object"}),
    )`
The Ubuntu/Data manager will pass, but other 21 checks will fail since the code
`pd.concat([df_time, df_float.iloc[:0]]) `
may have data type "datetime64" in these 21 checks(I only checked some).

File: pandas/tests/arithmetic/test_datetime64.py [hunk 1]
```diff
@@ -2436,3 +2436,34 @@ def test_dt64arr_addsub_object_dtype_2d():
...
     assert result2.shape == (4, 1)
     assert all(td._value == 0 for td in result2.ravel())
+
+
+def test_concat_float_datetime64():
+    # GH#32934
+    df_time = pd.DataFrame({"A": pd.array(["2000"], dtype="datetime64[ns]")})
+    df_float = pd.DataFrame({"A": pd.array([1.0], dtype="float64")})
+
+    tm.assert_equal(
```
Comment (mroeschke) [inline pandas/tests/arithmetic/test_datetime64.py:2446]: For each of these:

1. Could you use `tm.assert_frame_equal`
2. Create variables `expected = ... , result = ...`

**Comment by mroeschke** [comment]
Looks like the `using_array_manager` fixture needs to be used to set the dtype for the last example

File: pandas/tests/arithmetic/test_datetime64.py [hunk 2]
```diff
@@ -2436,3 +2436,40 @@ def test_dt64arr_addsub_object_dtype_2d():
 
     assert result2.shape == (4, 1)
     assert all(td._value == 0 for td in result2.ravel())
+
+
+def test_concat_float_datetime64(using_array_manager):
```
Comment (mroeschke) [inline pandas/tests/arithmetic/test_datetime64.py:2441]: Could you move this test to `pandas/tests/reshape/concat/test_datetimes.py`? After that this looks good

**Comment by liang3zy22** [comment]
It seems that another numpy error in "Ubuntu / Numpy Dev(pull_request)".
```
>       warnings.warn("`product` is deprecated as of NumPy 1.25.0, and will be "
[134](https://github.com/pandas-dev/pandas/actions/runs/4379948504/jobs/7666450861#step:8:136)
                      "removed in NumPy 2.0. Please use `prod` instead.",
[135](https://github.com/pandas-dev/pandas/actions/runs/4379948504/jobs/7666450861#step:8:137)
                      DeprecationWarning, stacklevel=2)
[136](https://github.com/pandas-dev/pandas/actions/runs/4379948504/jobs/7666450861#step:8:138)
E       DeprecationWarning: `product` is deprecated as of NumPy 1.25.0, and will be removed in NumPy 2.0. Please use `prod` instead.
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @liang3zy22 (that failure is unrelated)

**Pull request merged**